doc/source/reference/arrays.promotion.rst

```diff
@@ -6,90 +6,104 @@
+   result will have a *higher* precision than either inputs. For instance,
+   the result of an operation involving ``int64`` and ``uint64`` is
+   ``float64``.
 
 Please see the `Numerical promotion` section and image below for details
 on both.
 
 Detailed behavior of Python scalars
 -----------------------------------
-Since NumPy 2.0,[#NEP50]_ an important case in our promotion rules is that while
-NumPy usually never loses precision, it explicitly allows this for the Python
-numerical scalars of type ``int``, ``float``, and ``complex``.
-
-Python integers have arbitrary precision, and float and complex numbers
-have the same precision as NumPy's `float64` and `complex128`.
-However, unlike NumPy arrays and scalars, they do not have an explicit
-``dtype`` attached.  Because of this, NumPy assigns them the "kind" but
-ignores their actual precision when working with arrays of a lower precision
-dtype you do not expect this to change the result:
+Since NumPy 2.0 [#NEP50]_, an important point in our promotion rules is
+that although operations involving two NumPy dtypes never lose precision,
+operations involving a NumPy dtype and a Python scalar (``int``, ``float``,
+or ``complex``) *can* lose precision. For instance, it is probably intuitive
+that the result of an operation between a Python integer and a NumPy integer
+should be a NumPy integer. However, Python integers have arbitrary precision
+whereas all NumPy dtypes have fixed precision, so the arbitrary precision
+of Python integers cannot be preserved.
+
+More generally, NumPy considers the "kind" of Python scalars, but ignores
+their precision when determining the result dtype. This is often convenient.
+For instance, when working with arrays of a low precision dtype, it is usually
+desirable for simple operations with Python scalars to preserve the dtype.
 
   >>> arr_float32 = np.array([1, 2.5, 2.1], dtype="float32")
-  >>> arr_float32 + 10.0
+  >>> arr_float32 + 10.0  # undesirable to promote to float64
   array([11. , 12.5, 12.1], dtype=float32)
   >>> arr_int16 = np.array([3, 5, 7], dtype="int16")
-  >>> arr_int16 + 10
+  >>> arr_int16 + 10  # undesirable to promote to int64
   array([13, 15, 17], dtype=int16)
 
-In both cases the result precision is dictated only by the NumPy dtype.
-And because of that, the ``arr_float32 + 3.0`` behaves the same as
-``arr_float32 + np.float32(3.0)`` and ``arr_int16 + 10`` is found via
+In both cases, the result precision is dictated by the NumPy dtype.
+Because of this, ``arr_float32 + 3.0`` behaves the same as
+``arr_float32 + np.float32(3.0)``, and ``arr_int16 + 10`` behaves as
 ``arr_int16 + np.int16(10.)``.
-Although, when mixing NumPy integers and Python ``float`` or ``complex``
-the result always gives the default ``float64`` or ``complex128``:
+
+As another example, when mixing NumPy integers with a Python ``float``
+or ``complex``, the result always has type ``float64`` or ``complex128``:
```

Comment:
This doesn't quite follow from the rule of "NumPy assigns them the "kind" but ignores their actual precision..."

Reply:
Thanks for this!  Hmm, you are right.  The "kind" boundaries behave special in that way, not sure how to make a note of that well.

Will have a read through tomorrow and probably just adopt everyhing (and just push more changes if I want to tweak something).

Reply:
I guess you mean the fact that the important "kind" distinction is really more the integer vs inexact types.  Since unsigned/signed and float/complex could also be seen as the same kind but different range of values that can be represented.

But that seems even more complicated to point out, listing unsigned/signed integers as "kinds" makes it a bit easier to use the bit-sized precision...